src/libpmemfile-posix/data.c

```diff
@@ -0,0 +1,1604 @@
+/*
+ * Copyright 2016-2017, Intel Corporation
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in
+ *       the documentation and/or other materials provided with the
+ *       distribution.
+ *
+ *     * Neither the name of the copyright holder nor the names of its
+ *       contributors may be used to endorse or promote products derived
+ *       from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#define _GNU_SOURCE
+
+#include <limits.h>
+
+#include "callbacks.h"
+#include "data.h"
+#include "inode.h"
+#include "internal.h"
+#include "locks.h"
+#include "out.h"
+#include "pool.h"
+#include "os_thread.h"
+#include "util.h"
+#include "valgrind_internal.h"
+#include "ctree.h"
+
+static void
+expand_to_full_pages(uint64_t *offset, uint64_t *length)
+{
+	/* align the offset */
+	*length += *offset % FILE_PAGE_SIZE;
+	*offset -= *offset % FILE_PAGE_SIZE;
+
+	/* align the length */
+	*length = page_roundup(*length);
+}
+
+static void
+narrow_to_full_pages(uint64_t *offset, uint64_t *length)
+{
+	uint64_t end = page_rounddown(*offset + *length);
+	*offset = page_roundup(*offset);
+	if (end > *offset)
+		*length = end - *offset;
+	else
+		*length = 0;
+}
+
+/*
+ * block_cache_insert_block -- inserts block into the tree
+ */
+static void
+block_cache_insert_block(struct ctree *c, struct pmemfile_block *block)
+{
+	ctree_insert_unlocked(c, block->offset, (uintptr_t)block);
```

Comment:
Handling it can result in a lot of `if (stuff() != 0) return -1;`
How about adding a thread_local jmp_buf for these? If it would always be called from a transaction, I would go for an abort, but it is not always called from one.